src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+inline bool AnyQuery<T, N>::run(
+    const VectorType&       query_point,
+    const ValueType         query_max_square_distance
+#ifdef FOUNDATION_KNN_ENABLE_QUERY_STATS
+    , QueryStatistics&      stats
+#endif
+    ) const
+{
+    assert(!m_tree.empty());
+
+    FOUNDATION_KNN_QUERY_STATS(++stats.m_query_count);
+    FOUNDATION_KNN_QUERY_STATS(size_t fetched_node_count = 0);
+    FOUNDATION_KNN_QUERY_STATS(size_t visited_leaf_count = 0);
+    FOUNDATION_KNN_QUERY_STATS(size_t tested_point_count = 0);
+
+    const VectorType* APPLESEED_RESTRICT points = &m_tree.m_points.front();
+    const NodeType* APPLESEED_RESTRICT nodes = &m_tree.m_nodes.front();
+    const NodeType* APPLESEED_RESTRICT node;
+
+    const size_t IdealLeafSize = 40;
+
+    //
+    // Step 1:
+    //
+    //   Find the deepest node of the tree (leaf node or interior node) that contains
+    //   the query point and enough other points to compute a maximum search distance.
+    //   This descent is very fast since no node is pushed to the stack, and it allows
+    //   us to compute an initial maximum search distance which that speed up the real
+    //   search later on.
+    //
+
+    // Start at the root.
+    node = nodes;
+
+    FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+    // In any case, we must stop when we reach a leaf node.
+    while (node->is_interior())
+    {
+        const size_t split_dim = node->get_split_dim();
+        const ValueType split_abs = node->get_split_abs();
+        const ValueType split_dist = query_point[split_dim] - split_abs;
+
+        // Figure out which child node to follow (the one that contains the query point).
+        // Points on the split plane belong to the right child node.
+        const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+        const NodeType* APPLESEED_RESTRICT follow_node = nodes + node->get_child_node_index() + follow_index;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        // Stop at the parent node if the child node contains too few points.
+        if (follow_node->get_point_count() < IdealLeafSize)
+            break;
+
+        // Continue with the child node.
+        node = follow_node;
+    }
+
+    //
+    // Step 2:
+    //
+    //   Check if one of the points in this node (if it's a leaf node) or below this
+    //   node (if it's an interior node) is within the query distance.
+    //
+
+    {
+        FOUNDATION_KNN_QUERY_STATS(++visited_leaf_count);
+
+        const VectorType* APPLESEED_RESTRICT point_ptr = points + node->get_point_index();
+        const VectorType* APPLESEED_RESTRICT point_end = point_ptr + node->get_point_count();
+
+        while (point_ptr < point_end)
+        {
+            FOUNDATION_KNN_QUERY_STATS(++tested_point_count);
+
+            const ValueType square_dist = square_distance(*point_ptr++, query_point);
+
+            if (square_dist <= query_max_square_distance)
+                goto found;
```

Comment:
`goto` statements are badly to read as well. Also, the CI is complaining about it as well.
Maybe split the `run` function into several inlined subfunctions and make this a function call?

Reply:
Yeah I've removed the `goto`. I use them extremely rarely, but they occasionally come handy in performance critical code when needing to exit nested loops. But since CI complained I went for a cleaner solution.